js/widgets/widgetWindows.js

```diff
@@ -157,234 +124,396 @@ class WidgetWindow {
-            this._frame.remove();
+    /**
+     * @public
+     * @param {number} index 
+     * @param {string} icon 
+     * @param {number} iconSize 
+     * @param {string} label 
+     * @returns {HTMLElement}
+     */
+    modifyButton(index, icon, iconSize, label) {
+        this._buttons[index].innerHTML =
+            '<img src="header-icons/' +
+            icon +
+            '" title="' +
+            label +
+            '" alt="' +
+            label +
+            '" height="' +
+            iconSize +
+            '" width="' +
+            iconSize +
+            '" />';
+        return this._buttons[index];
+    }
 
-            window.widgetWindows.openWindows[this._key] = undefined;
-        };
+    /**
+     * @public
+     * @returns {void}
+     */
+    close() {
+        this.onclose();
+    }
 
-        this.onmaximize = () => {
-            return this;
-        };
+    /**
+     * The title may change, as with the Help Widget.
+     * @public
+     * @returns {void}
+     */
+    updateTitle(title) {
+        const wftTitle = docById(this._key + "WidgetID");
+        wftTitle.innerHTML = title;
+    }
 
-        this.show = () => {
-            this._frame.style.display = "block";
-        };
+    /**
+     * @public
+     * @returns {void}
+     */
+    takeFocus() {
+        const windows = docById("floatingWindows");
+        const siblings = windows.children;
+        for (let i = 0; i < siblings.length; i++) {
+            siblings[i].style.zIndex = "0";
+            siblings[i].style.opacity = ".7";
+        }
+        this._frame.style.zIndex = "1";
+        this._frame.style.opacity = "1";
+    }
 
-        this.setPosition = (x, y) => {
-            this._frame.style.left = x + "px";
-            this._frame.style.top = Math.max(y, 64) + "px";
-            window.widgetWindows._posCache[this._key] = [x, Math.max(y, 64)];
+    /**
+     * @public
+     * @param {*} icon 
+     * @param {*} iconSize 
+     * @param {*} label 
+     * @param {HTMLElement} parent
+     * @returns {HTMLElement} 
+     */
+    addButton(icon, iconSize, label, parent) {
+        console.log("From WidgetWindow addButton");
+        console.log(`type of icon = ${typeof(icon)}`);
+        console.log(`type of iconSize = ${typeof(iconSize)}`);
+        console.log(`type of label = ${typeof(label)}`);
+
+        const el = this._create("div", "wfbtItem", parent || this._toolbar);
+        el.innerHTML =
+            '<img src="header-icons/' +
+            icon +
+            '" title="' +
+            label +
+            '" alt="' +
+            label +
+            '" height="' +
+            iconSize +
+            '" width="' +
+            iconSize +
+            '" />';
+        this._buttons.push(el);
+        return el;
+    }
 
+    /**
+     * @public
+     * @returns {WidgetWindow} this
+     */
+    sendToCenter() {
+        const canvas = docById("myCanvas");
+        const fRect = this._frame.getBoundingClientRect();
```

Comment:
```node
Uncaught TypeError: Cannot read property 'getBoundingClientRect' of undefined
    at sendToCenter (widgetWindows.js:359)
```

Reply:
The reason for this error was how the sendToCenter function was being used in the code elsewhere. Which led to wrong value of `this` in it. Binding the context in the constructor solved this issue.